### ISSUE TITLE
[OS - Windows] Failed on build references project library - GetReferenceNearestTargetFrameworkTask not found

### ISSUE THREAD

**Issue by Hitmasu** [mono/monodevelop#4119]
Hi,

When i try build a library project who contains another library project as reference, occurs follow error:

`C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(5,5): Error MSB4036: The "GetReferenceNearestTargetFrameworkTask" task was not found. Check the following: 1.) The name of the task in the project file is the same as the name of the task class. 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface. 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin" directory. (MSB4036) (Project)`

The steps to repeat error:

1. Create a project library
2. Create another project library
3. Reference the project library **1** into **2** and try build project library 2.

I tried in different projects, builds and releases from MonoDevelop , but the error always happen.

****_That error not occurs in Visual Studio 2017 or in MSBuild.exe (Build command line)._**

> VS bug [#665484](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/665484)

**Comment by knocte** [comment]
> C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets

This is a windows path, are you sure you're running this in MonoDevelop? Where did you get your `MonoDevelop for Windows` binary from?

**Comment by Hitmasu** [comment]
Yes, i'm running latest version of MonoDevelop.

![monodevelop_2018-03-18_23-55-33](https://user-images.githubusercontent.com/13804176/37575825-077af5fc-2b08-11e8-8a6f-164214b9dc9a.png)

![monodevelop_2018-03-19_00-01-45](https://user-images.githubusercontent.com/13804176/37576000-bd8c884c-2b08-11e8-91f6-2506d933bfe4.png)

I follow that steps to build MonoDevelop on Windows: https://github.com/mono/md-website/blob/gh-pages/developers/building-monodevelop.md#prerequisites-and-source.

I tried in two different computer, one with only Build Tools 2017 installed and another with Visual Studio 2017, both return that error.

**Comment by knocte** [comment]
Ok cool, can you rename the title of this issue to make it clearer this is related to "Windows".

**Comment by Hitmasu** [comment]
Sorry, my bad.

**Comment by Hitmasu** [comment]
That error is not exclusive from MonoDevelop, i tried build with Mono Command Line and i get same error.

**Comment by knocte** [comment]
Ok I guess you need to move this to github.com/mono/mono then

**Comment by Hitmasu** [comment]
After long time, i finally found a way to fixed this. 

The problem is because is missing one file from **Nuget** : **Nuget.targets**.

When **MonoDevelop.Projects.Formats.MSBuild** copy files to `%AppData%\Local\MonoDevelop\7.0\Cache\MSBuild\(Random folder)`, for some reason, the file Nuget.targets was not copied. 

To fix this, i copied my NuGet.targets from 

`
C:\Program Files (x86)\Microsoft Visual Studio\2017\(Community | BuildTools )\Common7\IDE\CommonExtensions\Microsoft\NuGet `

To my MSBuildPath 

`
C:\Program Files (x86)\Microsoft Visual Studio\2017\(Community | BuildTools)\MSBuild\15.0\Bin`

Now, i can build with success my projects.

**Comment by KirillOsenkov** [comment]
We need to fix this properly.

**Comment by KirillOsenkov** [comment]
```
The "GetReferenceNearestTargetFrameworkTask" task was not found. Check the following: 1.) The name of the task in the project file is the same as the name of the task class. 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface. 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin" directory.
```

```xml
  <!-- Import NuGet.targets (required for GetReferenceNearestTargetFrameworkTask and used for Restore functionality) -->
  <PropertyGroup>
    <NuGetRestoreTargets Condition="'$(NuGetRestoreTargets)'=='' and '$([MSBuild]::IsRunningFromVisualStudio())'=='true'">$(MSBuildToolsPath32)\..\..\..\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets</NuGetRestoreTargets>
    <NuGetRestoreTargets Condition="'$(NuGetRestoreTargets)'==''">$(MSBuildToolsPath)\NuGet.targets</NuGetRestoreTargets>
  </PropertyGroup>
```

```xml
    <!-- For each reference, get closest match -->
    <GetReferenceNearestTargetFrameworkTask AnnotatedProjectReferences="@(_ProjectReferenceTargetFrameworkPossibilities)"
                                            CurrentProjectTargetFramework="$(ReferringTargetFrameworkForProjectReferences)"
                                            CurrentProjectName="$(MSBuildProjectName)"
                                            Condition="'@(_ProjectReferenceTargetFrameworkPossibilities->Count())' != '0' and '$(ReferringTargetFrameworkForProjectReferences)' != ''">
      <Output ItemName="AnnotatedProjects" TaskParameter="AssignedProjects" />
    </GetReferenceNearestTargetFrameworkTask>
```

**Comment by KirillOsenkov** [comment]
See related: https://github.com/Microsoft/msbuild/issues/2753

**Comment by KirillOsenkov** [comment]
Also related: https://github.com/Microsoft/msbuild/pull/2663

**Comment by KirillOsenkov** [comment]
PR out: https://github.com/mono/monodevelop/pull/5707

**Comment by spchamp** [comment]
Should NuGet.targets be installed along with MSBuild, or Mono?

I'm seeing a similar issue, on Linux, using a local MSBuild -- the build configuration may be a little complex to explain, in some few words. My local MSBuild installation was built and installed from the Mono MSBuild source distribution version 0.08, using a local (work-in-progress) port for pkgsrc, the Mono xplat-master binary distribution, and a Mono build (local) built from the latest Mono source code. I look forward to sharing the port information (build configuration) for these builds - some of this, comprising updates for pkgsrc wip ports - pursuant of further testing (LLVM integration, and furthermore). To the present issue, however....

My `/usr/pkg/lib/mono/msbuild/15.0/bin/Microsoft.Common.CurrentVersion.targets` file includes the directive for "Import NuGet.targets." 

I'm not sure if the `NuGet.targets` file is assumed to be provided by the project source code, or if it should be provided by the tooling framework. Perhaps, if I may be able to install that file -- in some ways -- correctly, it might serve to resolve the similar build failure that I'm seeing with mono-addins.

I'd found this issue page, while searching for more information about that build failure, to see if there may be any known workarounds for it. This is with a build on Linux (Mono and Mono MSBuild) such that -- I believe -- may be built from some latest source code.

I don't see any NuGet.targets file installed from Mono or from the Mono MSBuild distribution.  I'm not sure where the file is supposed to originate from, for the local Mono installation. 

I'll take another look at the MSBuild source tree, perhaps there may be a similar file available somewhere there. If I may be able to locate and install that file under an appropriate pathname for the local MSBuild installation, perhaps it may serve to resolve the similar build failure that I'm seeing, locally, albeit when building mono-addins (I'm trying to build and install the external dependencies for MonoDevelop, separate to the main MonoDevelop source tree, for purpose of porting to Linux and BSD operating systems under pkgsrc)

**Comment by KirillOsenkov** [comment]
@radical

**Comment by spchamp** [comment]
After installing the NuGet.targets file locally, it seems to have fixed the build failure. 

I'm not sure how the file was not installed, in the local MSBuild build configuration - will take another look at my local ports configuration. 

There is some helpful information, above, certainly.

**Comment by spchamp** [comment]
After installing the `NuGet.targets` file - as apparently should have been installed with my local MSBuild build, though I'm afraid my scripting may have missed the file - it seems to have fixed that specific build issue.

 I'll take another look at my msbuild-mono port, to try to diagnose why the file was not installed under my local build configuration. Perhaps this may serve to resolve some other build issues, furthermore. Definitely, there is some useful information above.

**Comment by spchamp** [comment]
As a further update on my albeit orthogonal remark, here: In the initial port build configuration for MSBuild with Mono, I'd not built the `mono/build/update_bundled_bits.proj` project before building `MSBuild.sln`.  It seems that this may have had a side effect, in that a number of files - including `NuGet.targets` - were not installed for the resulting pkg. I've updated the port build configuration so as to build `update_bundled_bits.proj` before `MBuild.sln`, such as it seems would be done under the build scripting for Azure CI. 

I've updated my local build configuration. I'm rebuilding my local MSBuild installation presently. Hopefully the resulting build will be more usable, after this update, perhaps of a quality that may be ready to publish as a contribution for Mono support in pkgsrc.
